internal/protocol/connector.go

```diff
@@ -214,18 +220,18 @@ func (c *Connector) connectAttemptAll(ctx context.Context, log logging.Func) (*P
 			}
 			log(logging.Debug, "connected")
 			pc <- protocol
-		}(server, protocolChan)
+		}(server, protocolCh)
 	}
 
 	// Read from protocol chan, cancel context
-	protocol, ok := <-protocolChan
+	protocol, ok := <-protocolCh
 	if !ok {
 		return nil, ErrNoAvailableLeader
 	}
 
 	cancel()
 
-	for extra := range protocolChan {
+	for extra := range protocolCh {
 		extra.Close()
 	}
```

Comment:
Just a question, with these changes, since we are not cancelling the dial context, does that mean `protocolCh` might still end up blocking this function from returning if a particular node is very slow and happens to have a request in flight?

If so, then I wonder if this loop should be in a goroutine, so that the leader is returned immediately before the remaining connections are closed? But that might cause spikes if there's repeated attempts to fetch the leader back-to-back.

Alternatively there could be another goroutine situated right after the call to `cancel()` which sleeps for some (configurable?) abort limit and then cancels the `origCtx` to potentially unblock `protocolCh`.  At least in my test setup about 100ms is fine enough to get past the TLS handshake, but that's not a very real-world test.

Reply:
> Just a question, with these changes, since we are not cancelling the dial context, does that mean protocolCh might still end up blocking this function from returning if a particular node is very slow and happens to have a request in flight?

Yep, that's true! I feel like maybe this is an argument for keeping a pool of open connections to the other nodes in the cluster instead of creating a bunch of new ones every time and then throwing them away. This requires dealing with membership changes, but we have a natural place to update the pool, namely the node store refresh hook. Hopefully this would also deal with the excessive CPU usage issues. What do you think?